### PR TITLE
feat(next): codegenDir and update .astro paths

### PR DESCRIPTION
**Pull request by florian-lefebvre**
## Changes

- Moves dts from `.astro/astro` to `.astro`. My original plan in 4.14 was to move stuff from `.astro` to `.astro/astro`. I did that partially back then (for content, env and actions) but I revert it in this PR (see comments)
- Adds `createCodegenDir()` to `astro:config:setup`. Points to `.astro/integrations/<normalized_integration_name>` after being called. This is convenient for integrations authors to be able to put files in a folder where they're sure they can't override files generated by Astro (eg. `.astro/types.d.ts`), nor files generated by another integration. In 4.14, we introduced `injectTypes` which writes files to this `codegenDir` directory already, so we just expose it now

## Testing

Should still pass

## Docs

Changeset + https://github.com/withastro/docs/pull/9495

### CODE REVIEW

**Comment by Princesseuh** [comment]
For language-tools this mean that we need to load from multiple paths, here https://github.com/withastro/language-tools/blob/b79ff829224838492fcdfa273a03316863e724da/packages/ts-plugin/src/index.ts#L12 and here https://github.com/withastro/language-tools/blob/b79ff829224838492fcdfa273a03316863e724da/packages/language-server/src/nodeServer.ts#L45

**Comment by florian-lefebvre** [comment]
Done in https://github.com/withastro/language-tools/pull/951

**Comment by bluwy** [comment]
Why are we moving all the files one-level down? Don't we control `.astro` so we are free to structure it however we want? Is this making way for other integrations to put stuff in `.astro` without naming collisions? If that's true, I don't think we should be explicitly catering it as they should be using a unique enough name (like `.astro/my-integration-special-data.json`) to prevent collision chances.

**Comment by florian-lefebvre** [comment]
ATM we can't control where people put stuff. I'm in favor of creating somle kind of convention (`.astro/astro` for core) there. At least that was the intent when doing `injectTypes` in 4.14. To protect ourselves even more, we'd need a `injectAsset` (probably bad name) that would do basically the same as `injectTypes` BUT for any file and without creating references.

**Comment by bluwy** [comment]
Do we want to say that it's safe to put your stuff in `.astro`? There's always a risk borrowing a folder you don't own and putting stuff there, and I think it should apply here too.

Astro owns the `.astro` folder so I don't think we need to protect ourselves from integrations tampering it. If an integration caused a name conflict in the `.astro` folder or something else that caused Astro to crash, the integration should be fixing it, not us.

For `injectTypes`, I see it more as a way for integrations to easily add types and sync with our command. And deduplicates the same amount of work to construct and write types.

**Comment by florian-lefebvre** [comment]
> There's always a risk borrowing a folder you don't own and putting stuff there, and I think it should apply here too

It makes sense to me, as an integration author, to want to put codegen files in there since that's where astro does it. otherwise, that means you need to create your own `.folder`, add it to gitignore etc

---

As for the rest, I get your point, I guess we could move things outside of `.astro/astro` instead. I'd like to at least expose the directory `.astro/integrations/<normalized>` to integrations authors so that they can put stuff in there without conflicting with astro. wdyt?

**Comment by bluwy** [comment]
I think it's fine for integrations to put stuff into `.astro`, just that we can't guarantee to not break your integration if we decide to add certain file names that could collide with yours. So if integrations want to avoid that risk, the names that they put in there should be unique enough. 

For `.astro/integrations/`, I'm fine with that if it's only documentation that you're encouraged to put your assets there if needed. I'm not so sure about codifying the guarantee or like helping generate the folder. Maybe others having different thoughts on this.

**Comment by florian-lefebvre** [comment]
Doesn't have to be in the beta so not urgent to answer. @Princesseuh when you have some time I'd like your opinion on that

**Comment by florian-lefebvre** [comment]
Alright then if you're both okay with that, I'll do the following:
- close the language tools PR
- move dts from `.astro/astro` to `.astro`
- expose the integration dir as URL in `astro:config:setup`

**Review by ematipico — DISMISSED** [review]
To be honest, with the little knowledge I have of `injectTypes` when I read this changeset, I failed to understand what `codegenDir` is for and how it is related to `injectTypes`.

For example, the following phrase doesn't say anything to me. I don't understand what space you're referring to, and how other integrations are involved in this feature.
> It allows you to have space without risking being overriden by another integration

I would try to help with the review, but the poor description of the PR and the changeset don't help.

Can I ask you to revisit the changeset, and make a ELI5 version of it?

**Comment by florian-lefebvre** [comment]
@ematipico I clarified both PR desc and changeset, thanks for the feedback!

**Review by ematipico — COMMENTED** [review]
Here's a suggestion. In the changeset you highlight the fact that integrations now don't conflict each, and the hook provide the reserved directory to do so.

However, nothing blocks users from doing `new URL("../other-integration", codegenDir)`, and that's how integrations can tamper with each other. Sure, integrations shouldn't do that.

Instead, I think we should be stricter and provide callbacks so we have finer control, for example:

```js
const integration = {
    name: 'my-integration',
    hooks: {
        'astro:config:setup': ({ writeToCodegenDir, purgeCodegenDir, readFromCodegeDir }) => {
						purgeCodegenDir();
						writeToCodegenDir(() => {
							return { filename: "file.ts", contents: "{}" }
						});
						const map = await readFromCodegeDir();
						for (const [filePath, contents] of map) {

						}
        }
    }
}
```

File: packages/astro/src/integrations/hooks.ts [hunk 1]
```diff
@@ -151,6 +155,9 @@ export async function runHookConfigSetup({
 	for (let i = 0; i < updatedConfig.integrations.length; i++) {
 		const integration = updatedConfig.integrations[i];
 
+		const codegenDir = new URL(normalizeCodegenDir(integration.name), settings.dotAstroDir);
+		await fs.promises.mkdir(codegenDir, { recursive: true });
```
Comment (matthewp) [inline packages/astro/src/integrations/hooks.ts:159]: Does this mean we're making a directory for every integration even if they are not using it? Assuming so, let's talk about that. It feels like a lot initially, I could be overthinking this though.

An alternative would be something like a `createCodegenDir()` that returns the path, so the integration explicitly asks for it.

Comment (florian-lefebvre) [reply packages/astro/src/integrations/hooks.ts:159]: That sounds good to me

File: .changeset/three-olives-reflect.md [hunk 2]
```diff
@@ -2,21 +2,22 @@
...
 
 While `.astro` has always been the preferred place to write code generated files, it has also been prone to mistakes. For example, you can write a `.astro/types.d.ts` file, breaking Astro types. Or you can create a file that overrides a file created by another integration.
 
-In this release, `<root>/.astro/integrations/<normalized_integration_name>` is now exposed in the `astro:config:setup` hook as `codegenDir`. It allows you to have a dedicated folder, avoiding conflicts with another integration or Astro. This directory is always created before any hook runs so it's safe to write files to it directly:
+In this release, `<root>/.astro/integrations/<normalized_integration_name>` can now be retrieved in the `astro:config:setup` hook by calling `createCodegenDir()`. It allows you to have a dedicated folder, avoiding conflicts with another integration or Astro. This directory is created by calling thos function so it's safe to write files to it directly:
```
Comment (Princesseuh) [inline .changeset/three-olives-reflect.md:11]:
```suggestion
In this release, `<root>/.astro/integrations/<normalized_integration_name>` can now be retrieved in the `astro:config:setup` hook by calling `createCodegenDir()`. It allows you to have a dedicated folder, avoiding conflicts with another integration or Astro. This directory is created by calling this function so it's safe to write files to it directly:
```

**Review by bluwy — APPROVED** [review]

**Review by sarah11918 — APPROVED** [review]
Thanks @florian-lefebvre ! Just some tiny language nits here, but looks great!
